File: skills/adkit/scripts/src/ideas/urls.ts
/**
 * Final-URL collection + reachability for the /adkit create scaffold.
 *
 * `finalUrls(brief)` is pure (no network) — it just enumerates the brief's
 * destination URLs, so it is unit-testable without touching the wire. The actual
 * HEAD/GET probe lives in `urlUnreachableReason` (its natural home), and
 * `unreachableUrls` composes the two: collect, probe each, return failures.
 *
 * No stdout, no sys.exit — bin/create formats and dies on the failure list.
 */

import type { Brief } from "../lib/schema.js";

/**
 * Every destination URL the brief publishes: one per RSA + one per sitelink.
 * Deduped, order-preserving.
 */
export function finalUrls(brief: Brief): string[] {
  const rsaUrls = brief.adGroups.map((ag) => String(ag.responsiveSearchAd.finalUrl));
  const sitelinkUrls = brief.campaign.sitelinks.map((sl) => String(sl.finalUrl));
  return [...new Set([...rsaUrls, ...sitelinkUrls])];
}

/**
 * `null` if the URL resolves (status < 400, redirects followed); else a short
 * reason. HEAD first, fall back to GET when the host rejects HEAD.
 */
export async function urlUnreachableReason(url: string): Promise<string | null> {
  for (const method of ["HEAD", "GET"] as const) {
    try {
      const resp = await fetch(url, {
        method,
        redirect: "follow",
        headers: { "User-Agent": "ads-skill-urlcheck" },
        signal: AbortSignal.timeout(10_000),
      });
      // Some servers reject HEAD with 403/405 — retry with GET.
      if ((resp.status === 403 || resp.status === 405) && method === "HEAD") {
        continue;
      }
      return resp.status < 400 ? null : `HTTP ${resp.status}`;
    } catch (exc) {
      // DNS, timeout, TLS, etc. — surface the error's type name.
      return exc instanceof Error ? exc.name : String(exc);
    }
  }
  return null;
}

/**
 * The brief's destination URLs that don't resolve, as [url, reason] pairs.
 * Empty when every URL is reachable. Catches the classic /ideas/ prefix slip
 * and leftover TODO slugs before any Google Ads mutation runs.
 */
export async function unreachableUrls(brief: Brief): Promise<Array<[string, string]>> {
  const probed = await Promise.all(
    finalUrls(brief).map(async (url): Promise<[string, string | null]> => [url, await urlUnreachableReason(url)]),
  );
  return probed.filter((pair): pair is [string, string] => pair[1] !== null);
}

File: skills/adkit/scripts/ads_skill/ideas/urls.py
"""Final-URL collection + reachability for the /adkit create scaffold.

`final_urls(brief)` is pure (no network) — it just enumerates the brief's
destination URLs, so it is unit-testable without touching the wire. The actual
HEAD/GET probe lives in `url_unreachable_reason` (its natural home), and
`unreachable_urls` composes the two: collect, probe each, return failures.

No stdout, no sys.exit — bin/create.py formats and dies on the failure list."""

from __future__ import annotations

import urllib.error
import urllib.request

from ..lib.schema import Brief


def final_urls(brief: Brief) -> list[str]:
    """Every destination URL the brief publishes: one per RSA + one per sitelink.
    Deduped, order-preserving."""
    urls = [str(ag.responsiveSearchAd.finalUrl) for ag in brief.adGroups]
    urls += [str(sl.finalUrl) for sl in brief.campaign.sitelinks]
    return list(dict.fromkeys(urls))


def url_unreachable_reason(url: str) -> str | None:
    """None if the URL resolves (status < 400, redirects followed); else a short reason.
    HEAD first, fall back to GET when the host rejects HEAD."""
    for method in ("HEAD", "GET"):
        req = urllib.request.Request(url, method=method, headers={"User-Agent": "ads-skill-urlcheck"})
        try:
            with urllib.request.urlopen(req, timeout=10) as resp:
                return None if resp.status < 400 else f"HTTP {resp.status}"
        except urllib.error.HTTPError as exc:
            if exc.code in (403, 405) and method == "HEAD":
                continue  # some servers reject HEAD — retry with GET
            return f"HTTP {exc.code}"
        except Exception as exc:  # noqa: BLE001 — DNS, timeout, TLS, etc.
            return type(exc).__name__
    return None


def unreachable_urls(brief: Brief) -> list[tuple[str, str]]:
    """The brief's destination URLs that don't resolve, as (url, reason) pairs.
    Empty when every URL is reachable. Catches the classic /ideas/ prefix slip
    and leftover TODO slugs before any Google Ads mutation runs."""
    return [
        (url, reason)
        for url in final_urls(brief)
        if (reason := url_unreachable_reason(url)) is not None
    ]
